galsim/chromatic.py

```diff
@@ -2005,6 +2032,101 @@ def noise(self):
         return BaseCorrelatedNoise(noise.rng, noise_prof, noise.wcs)
 
 
+class SimpleChromaticTransformation(ChromaticTransformation):
+    """A class for the simplest kind of chromatic object -- a GSObject times an SED.
+
+    This is a subclass of ChromaticTransformation, which just skips some calculations
+    that are unnecessary in this simple, but fairly common special case.
+
+    Parameters:
+        obj:                The object to be transformed.
+        sed:                An `SED` instance.
+        gsparams:           An optional `GSParams` argument.  See the docstring for `GSParams` for
+                            details. [default: None]
+        propagate_gsparams: Whether to propagate gsparams to each of the components.  This
+                            is normally a good idea, but there may be use cases where one
+                            would not want to do this. [default: True]
+    """
+    def __init__(self, obj, sed=1., gsparams=None, propagate_gsparams=True):
+        self.chromatic = False
+        self.separable = True
+        self.interpolated = False
+        self._redshift = None
+
+        self._gsparams = GSParams.check(gsparams, obj.gsparams)
+        self._propagate_gsparams = propagate_gsparams
+
+        self._original = obj
+        self._jac = None
+        self._offset = (0.,0.)
+        self._flux_ratio = sed
+
+        if self._propagate_gsparams:
+            self._original = self._original.withGSParams(self._gsparams)
+        self.deinterpolated = self
+
+    @property
+    def sed(self):
+        return self._flux_ratio * self.original.flux
+
+    def __hash__(self):
+        if not hasattr(self, '_hash'):
+            self._hash = hash(("galsim.SimpleChromaticTransformation", self.original,
+                               self._flux_ratio, self._gsparams, self._propagate_gsparams))
+        return self._hash
+
+    def __repr__(self):
+        return ('galsim.SimpleChromaticTransformation(%r, sed=%r, '
+                'gsparams=%r, propagate_gsparams=%r)')%(
+            self.original, self._flux_ratio,
+            self._gsparams, self._propagate_gsparams)
+
+    def __str__(self):
+        return str(self.original) + ' * ' + str(self.sed)
+
+    def _getTransformations(self, wave):
+        flux_ratio = self._flux_ratio(wave)
+        return self._jac, self._offset, flux_ratio
+
+    def _shoot(self, photons, rng):
```

Comment:
Hm.  Good point  I'm pretty sure this was just the normal ChromaticTransformation._shoot function stripped of the parts that weren't relevant.  So I think that one also has the wrong noise properties.  It's quite possible others do as well.

I'll see if I can think of a good unit test for this that we can run on all the chromatic objects to see which ones break the expected Poisson statistics.  It might not be possible to get it perfect for non-separable chromatic profiles.  But we can at least get it right for the separable ones.

In any case, the right implementation for this one is, as you say, to not rescale the fluxes, but just assign wavelengths the way WavelengthSampler does.

Reply:
So, I looked into this a bit, and this function only gets used for PSFs (i.e. dimensionless chromatic objects).  The base separable galaxy object with an SED does use the WavelengthSampler already.  But then the PSFs are applied as photon operators, which act on the photons as they come in.  So if one of them has a wavelength dependent flux scaling (not sure how that would arise physically, but we allow it), then it scales the flux values of the photons as requested.  I *think* this is probably right.  It's just a weird edge case that probably doesn't happen much in practice.

Reply:
Sounds good.  If this is only for PSFs (sounds right) then yeah, I don't see how this could ever even reasonably get triggered.  Even chromatic PSFs ought to have flux=1 at every wavelength I think.  Thanks for taking a look.